pkg/kapp/clusterapply/converged_resource.go

```diff
@@ -41,6 +70,35 @@ func (c ConvergedResource) IsDoneApplying() (ctlresm.DoneApplyState, []string, e
 		return ctlresm.DoneApplyState{Done: true}, descMsgs, err
 	}
 
+	// Custom wait rule
+	wr := c.waitingRule
+	if wr.SupportsObservedGeneration || len(wr.FailureConditions) > 0 || len(wr.SuccessfulConditions) > 0 {
+		obj := genericResource{}
+		err := c.res.AsUncheckedTypedObj(&obj)
+		if err != nil {
+			return ctlresm.DoneApplyState{Done: true, Successful: false}, descMsgs, err
+		}
+		if wr.SupportsObservedGeneration && obj.Metadata.Generation != obj.Status.ObservedGeneration {
+			return ctlresm.DoneApplyState{Done: false}, descMsgs, err
+		}
+		for _, fc := range wr.FailureConditions {
+			for _, cond := range obj.Status.Conditions {
+				if cond.Type == fc && cond.Status == v1.ConditionTrue {
+					descMsgs = append(descMsgs, cond.Message)
+					return ctlresm.DoneApplyState{Done: false, Successful: false}, descMsgs, err
```

Comment:
let's use Message field on DoneApplyState similar to example here: https://github.com/k14s/kapp/blob/master/pkg/kapp/resourcesmisc/core_v1_service.go#L30